File: lib/toolRun.ts
/**
 * Tool-run aggregation payload (protocol v10, bridge kind 6).
 *
 * The host owns the reachable `/api/agent` SSE and knows the structured
 * `tool_result.ok`, so it aggregates each uninterrupted tool streak into ONE
 * display-only transcript message (`MessageKind.ToolRun`, session role
 * `tool_run`). The Wasm harness decodes this versioned, delimiter-encoded
 * payload and paints an expandable "N tools called" control.
 *
 * Format (protocol v10 / TOOL_RUN_VERSION 1):
 *
 *   toolrun\t1\t{ok}/{fail}/{pending}     header: ok/fail/pending counts (total = sum)
 *   {id}\t{status}\t{name}\t{brief}\t{detail}
 *   ...
 *
 * `status ∈ running | ok | fail`. Fields `name`/`brief`/`detail` are escaped for
 * `\t`, `\n`, `\\`; decode fails open (returns null) on a bad header or unknown
 * version so the caller renders the raw body as plain text and never crashes.
 *
 * See also `native/harness/src/rich/toolrun.zig` (mirror decoder, host-tested).
 */

export const TOOL_RUN_VERSION = 1 as const;

/**
 * Max items stored in a single `tool_run` group. Keeps any one session message
 * well under the cloud `~262 KiB`/msg cap even for very long uninterrupted
 * agentic streaks; when a group would exceed this the host rolls a new group
 * (counts stay exact across groups).
 */
export const TOOL_RUN_ITEMS_MAX = 200 as const;

export type ToolRunStatus = 'running' | 'ok' | 'fail';

export interface ToolRunItem {
  /** 1-based id within the group — stable key for per-item expand state. */
  id: number;
  status: ToolRunStatus;
  name: string;
  /** One-liner level-1 preview (single line, ≤ BRIEF_PREVIEW_MAX chars). */
  brief: string;
  /** Full level-2 detail — the already-truncated per-tool summary (may be long). */
  detail: string;
}

/**
 * Level-1 preview cap. Level-2 `detail` keeps the full (already server-truncated)
 * tool summary; the collapsed `brief` is a collapse-whitespace preview so the
 * expand-to-detail tier genuinely adds the full text once a summary exceeds a
 * single short line. Empty/all-whitespace summaries fall back to a name+status
 * one-liner.
 */
export const BRIEF_PREVIEW_MAX = 64 as const;

function compactPreview(s: string): string {
  const t = (s ?? '').replace(/\s+/g, ' ').trim();
  if (!t) return '';
  return t.length > BRIEF_PREVIEW_MAX ? `${t.slice(0, BRIEF_PREVIEW_MAX)}…` : t;
}

/**
 * The colored status glyph column is the single symbol status channel for a
 * tool row (adversarial review #358 Major). The tool summary still embeds a
 * `✓`/`✗` mark (`name · ✓ ok · …`) — those code points are NOT in the Noto text
 * faces the L1 expander label and L2 detail body paint with (only the mark
 * column uses the DejaVu Sans Symbols face that covers them), so leaving them in
 * `brief`/`detail` rendered `.notdef` tofu right beside the real glyph. Rewrite
 * the deterministic `· ✓ ok` / `· ✗ failed` token to its ASCII letters only; the
 * symbol glyph lives in the mark column alone.
 */
export function asciiStatus(s: string): string {
  return s
    .replace(/·\s*✓\s*ok/g, '· ok')
    .replace(/·\s*✗\s*failed/g, '· failed');
}

/**
 * The L1 expander label is painted by dvui as a SINGLE Noto run — it has no
 * per-glyph symbol fallback the way the L2 body does (`mixed_text.addTextMixed`
 * routes symbols/emoji to a DejaVu symbols / OpenMoji face). So beyond the
 * `✓`/`✗` status marks, the collapsed `brief` must not carry any symbol the Noto
 * heading face lacks. The deterministic summary uses `→` (http_get lines), which
 * is NOT in Noto — map it to ASCII for the collapsed preview only. The expanded
 * L2 detail keeps the real glyph (rendered by the symbols-aware body).
 */
function briefSafe(s: string): string {
  return s.replace(/→/g, '->');
}

export type ToolRunCounts = {
  ok: number;
  fail: number;
  pending: number;
};

export interface ToolRunPayload {
  ok: number;
  fail: number;
  pending: number;
  items: ToolRunItem[];
}

/** Mutable in-turn aggregation state (host `harnessChat.runHarnessTurn`). */
export interface ToolRunGroup {
  items: ToolRunItem[];
}

export function createToolRunGroup(): ToolRunGroup {
  return { items: [] };
}

/** True once a group stores TOOL_RUN_ITEMS_MAX items (roll to a new group). */
export function toolRunIsFull(group: ToolRunGroup): boolean {
  return group.items.length >= TOOL_RUN_ITEMS_MAX;
}

/**
 * True when the group already has a `running` item with this name. Used to
 * decide whether a `tool_result` grows the group (no running match → new item).
 */
export function hasRunningTool(group: ToolRunGroup, name: string): boolean {
  for (let i = group.items.length - 1; i >= 0; i--) {
    const it = group.items[i];
    if (it && it.status === 'running' && it.name === name) return true;
  }
  return false;
}

export function addToolStart(group: ToolRunGroup, name: string): void {
  group.items.push({
    id: group.items.length + 1,
    status: 'running',
    name,
    brief: `${name} · running…`,
    detail: '',
  });
}

/**
 * Level-2 `detail` for a tool row. Prefers the bounded `preview` (real
 * command/output detail from the backend tool_result) when it is meaningfully
 * richer than the collapsible L1 `brief`; otherwise returns empty so the Wasm
 * paints a static label instead of a duplicate-of-L1 blank expander (phase 3
 * #353 / parent #352 decision C).
 */
export function meaningfulDetail(
  summary: string,
  preview: string | undefined,
): string {
  const line = asciiStatus((summary ?? '').trim());
  if (!line) return '';
  // Compare against the sanitized L1 brief (arrows already ASCII in brief) so a
  // `→`-bearing preview is not mistaken for the identical one-liner.
  const brief = briefSafe(compactPreview(line));
  const p = (preview ?? '').replace(/\r\n/g, '\n').trim();
  if (!p) return '';
  // Identical to the collapsed one-liner (or the full summary) → no pretend expand.
  if (p === brief || p === line) return '';
  return p;
}

export function addToolResult(
  group: ToolRunGroup,
  name: string,
  ok: boolean,
  summary: string,
  preview?: string,
): void {
  const line = asciiStatus((summary ?? '').trim());
  const detail = meaningfulDetail(summary, preview);
  // Complete the most recent running item with the same name (in place), else
  // append a done item (e.g. a result for a start we never observed).
  for (let i = group.items.length - 1; i >= 0; i--) {
    const it = group.items[i];
    if (it && it.status === 'running' && it.name === name) {
      it.status = ok ? 'ok' : 'fail';
      it.brief = briefSafe(compactPreview(line)) || `${name} · ${ok ? 'ok' : 'failed'}`;
      it.detail = detail;
      return;
    }
  }
  group.items.push({
    id: group.items.length + 1,
    status: ok ? 'ok' : 'fail',
    name,
    brief: briefSafe(compactPreview(line)) || `${name} · ${ok ? 'ok' : 'failed'}`,
    detail,
  });
}

export function countToolRunItems(items: ToolRunItem[]): ToolRunCounts {
  let ok = 0;
  let fail = 0;
  let pending = 0;
  for (const it of items) {
    if (it.status === 'ok') ok++;
    else if (it.status === 'fail') fail++;
    else pending++;
  }
  return { ok, fail, pending };
}

/**
 * Build tool-run groups from a non-stream `agentResult.toolTrace`, chunking by
 * `TOOL_RUN_ITEMS_MAX`. Void/empty → empty array.
 */
export function buildTraceGroups(
  trace: { name: string; ok: boolean; summary: string }[] | undefined,
): ToolRunGroup[] {
  if (!trace?.length) return [];
  const groups: ToolRunGroup[] = [];
  let cur = createToolRunGroup();
  for (const entry of trace) {
    const name = entry.name || 'tool';
    const line = asciiStatus((entry.summary ?? '').trim());
    const runningMatch = cur.items.find(
      (it) => it.status === 'running' && it.name === name,
    );
    if (!runningMatch && cur.items.length >= TOOL_RUN_ITEMS_MAX) {
      groups.push(cur);
      cur = createToolRunGroup();
    }
    if (runningMatch) {
      runningMatch.status = entry.ok ? 'ok' : 'fail';
      runningMatch.brief =
        briefSafe(compactPreview(line)) || `${name} · ${entry.ok ? 'ok' : 'failed'}`;
      runningMatch.detail = line;
    } else {
      cur.items.push({
        id: cur.items.length + 1,
        status: entry.ok ? 'ok' : 'fail',
        name,
        brief:
          briefSafe(compactPreview(line)) || `${name} · ${entry.ok ? 'ok' : 'failed'}`,
        detail: line,
      });
    }
  }
  if (cur.items.length > 0) groups.push(cur);
  return groups;
}

function esc(s: string): string {
  return s
    .replace(/\\/g, '\\\\')
    .replace(/\n/g, '\\n')
    .replace(/\t/g, '\\t');
}

function unesc(s: string): string {
  let out = '';
  for (let i = 0; i < s.length; i++) {
    const c = s[i];
    if (c === '\\' && i + 1 < s.length) {
      const n = s[i + 1];
      if (n === 't') {
        out += '\t';
        i++;
        continue;
      }
      if (n === 'n') {
        out += '\n';
        i++;
        continue;
      }
      if (n === '\\') {
        out += '\\';
        i++;
        continue;
      }
    }
    out += c;
  }
  return out;
}

/** Encode a non-empty group; returns null when there are no items. */
export function encodeToolRun(group: ToolRunGroup): string | null {
  if (group.items.length === 0) return null;
  const c = countToolRunItems(group.items);
  const lines: string[] = [
    `toolrun\t${TOOL_RUN_VERSION}\t${c.ok}/${c.fail}/${c.pending}`,
  ];
  for (const it of group.items) {
    lines.push(
      `${it.id}\t${it.status}\t${esc(it.name)}\t${esc(it.brief)}\t${esc(it.detail)}`,
    );
  }
  return lines.join('\n');
}

/**
 * Decode a tool-run payload. Returns null (fail-open) on a malformed header or
 * unknown version; tolerantly skips malformed item lines.
 */
export function decodeToolRun(text: string): ToolRunPayload | null {
  if (!text) return null;
  const lines = text.split('\n');
  if (lines.length === 0) return null;
  const head = lines[0].split('\t');
  if (head.length < 3) return null;
  if (head[0] !== 'toolrun') return null;
  if (Number(head[1]) !== TOOL_RUN_VERSION) return null;
  const counts = head[2].split('/');
  // Parsed for header validity only — recount from the kept items below so a
  // hostile/dense blob's header can never disagree with what actually decodes
  // after the TOOL_RUN_ITEMS_MAX cap (review Minor).
  if (
    counts.length !== 3 ||
    !counts.every((c) => c !== undefined && c !== '' && Number.isInteger(Number(c)))
  ) {
    return null;
  }
  const items: ToolRunItem[] = [];
  for (let i = 1; i < lines.length; i++) {
    // Defensive cap mirroring the host grouping bound (a restored blob could
    // carry more than TOOL_RUN_ITEMS_MAX; stop reading so a hostile/dense
    // payload can't force unbounded decode).
    if (items.length >= TOOL_RUN_ITEMS_MAX) break;
    const row = lines[i];
    if (!row) continue;
    const p = row.split('\t');
    if (p.length < 5) continue;
    const id = Number(p[0]);
    const status = p[1] as ToolRunStatus;
    if (status !== 'running' && status !== 'ok' && status !== 'fail') continue;
    items.push({
      id: Number.isInteger(id) && id > 0 ? id : i,
      status,
      name: unesc(p[2]),
      brief: unesc(p[3]),
      detail: unesc(p[4]),
    });
  }
  // Recount from the kept items so the header display can never disagree with
  // the decoded list after the TOOL_RUN_ITEMS_MAX cap (review Minor).
  const c = countToolRunItems(items);
  return { ok: c.ok, fail: c.fail, pending: c.pending, items };
}

File: native/harness/src/rich/toolrun.zig
//! Tool-run payload decoder (protocol v10, kind 6) — pure, host-testable.
//!
//! Mirror of `lib/toolRun.ts` (TS encoder). The host aggregates each
//! uninterrupted tool streak into one `tool_run` session message and encodes it
//! as versioned, tab/newline-delimited text:
//!
//!   toolrun\t1\t{ok}/{fail}/{pending}
//!   {id}\t{status}\t{name}\t{brief}\t{detail}
//!   ...
//!
//! `status ∈ running | ok | fail`. Fields `name`/`brief`/`detail` escape
//! `\t`, `\n`, `\\`. A malformed header or unknown `version` **fails open**
//! (returns null) so the caller renders the raw body as plain text and never
//! crashes. Malformed item lines are tolerated (skipped).
const std = @import("std");

pub const TOOL_RUN_VERSION: u8 = 1;
pub const MAX_ITEMS: u32 = 200;

pub const Status = enum(u8) {
    running = 0,
    ok = 1,
    fail = 2,
};

pub const Item = struct {
    id: u32,
    status: Status,
    name: []const u8,
    brief: []const u8,
    detail: []const u8,
};

pub const ToolRun = struct {
    ok: u32 = 0,
    fail: u32 = 0,
    pending: u32 = 0,
    items: []Item,
};

/// Owns the allocations backing a decoded `ToolRun` (string + item slices).
pub const Decoded = struct {
    run: ToolRun,
    alloc: std.mem.Allocator,

    pub fn deinit(self: *Decoded) void {
        for (self.run.items) |it| {
            self.alloc.free(it.name);
            self.alloc.free(it.brief);
            self.alloc.free(it.detail);
        }
        self.alloc.free(self.run.items);
    }
};

fn parseStatus(s: []const u8) ?Status {
    if (std.mem.eql(u8, s, "running")) return .running;
    if (std.mem.eql(u8, s, "ok")) return .ok;
    if (std.mem.eql(u8, s, "fail")) return .fail;
    return null;
}

fn parseIntOrZero(s: ?[]const u8) u32 {
    const v = s orelse return 0;
    return std.fmt.parseInt(u32, v, 10) catch 0;
}

/// Unescape `\t`→TAB, `\n`→LF, `\\`→backslash into a fresh buffer.
fn unescape(alloc: std.mem.Allocator, s: []const u8) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(alloc);
    var i: usize = 0;
    while (i < s.len) : (i += 1) {
        const c = s[i];
        if (c == '\\' and i + 1 < s.len) {
            const n = s[i + 1];
            if (n == 't') {
                try out.append(alloc, '\t');
                i += 1;
                continue;
            }
            if (n == 'n') {
                try out.append(alloc, '\n');
                i += 1;
                continue;
            }
            if (n == '\\') {
                try out.append(alloc, '\\');
                i += 1;
                continue;
            }
        }
        try out.append(alloc, c);
    }
    return try out.toOwnedSlice(alloc);
}

/// Decode a protocol v1 tool-run payload, allocating from `alloc` (use a
/// frame arena; call `Decoded.deinit` when the caller owns the lifetime).
/// Returns null (fail-open) on a malformed header or unknown version.
pub fn decode(alloc: std.mem.Allocator, text: []const u8) ?Decoded {
    var lines = std.mem.splitScalar(u8, text, '\n');
    const head = lines.next() orelse return null;
    var hf = std.mem.splitScalar(u8, head, '\t');
    const tag = hf.next() orelse return null;
    if (!std.mem.eql(u8, tag, "toolrun")) return null;
    const ver_s = hf.next() orelse return null;
    const ver = std.fmt.parseInt(u8, ver_s, 10) catch return null;
    if (ver != TOOL_RUN_VERSION) return null;
    const counts_s = hf.next() orelse return null;
    var cc = std.mem.splitScalar(u8, counts_s, '/');
    // Parsed for header validity only — we recount from the kept items below so
    // a hostile/dense blob's header can never disagree with what actually paints
    // (e.g. header "205/0/0" while the MAX_ITEMS cap keeps only 200 rows).
    _ = parseIntOrZero(cc.next());
    _ = parseIntOrZero(cc.next());
    _ = parseIntOrZero(cc.next());

    var items: std.ArrayList(Item) = .empty;
    errdefer items.deinit(alloc);
    // Defensive cap mirroring the host grouping bound (a restored/cloud/local
    // blob can carry more than MAX_ITEMS; stop so a hostile/dense payload can't
    // force unbounded per-item decode+unescape every frame).
    while (lines.next()) |lin| {
        if (items.items.len >= MAX_ITEMS) break;
        if (lin.len == 0) continue;
        var f = std.mem.splitScalar(u8, lin, '\t');
        const id_s = f.next() orelse continue;
        const status_s = f.next() orelse continue;
        const name_s = f.next() orelse continue;
        const brief_s = f.next() orelse continue;
        const detail_s = f.next() orelse continue;
        const id = std.fmt.parseInt(u32, id_s, 10) catch continue;
        const status = parseStatus(status_s) orelse continue;
        const name = unescape(alloc, name_s) catch continue;
        const brief = unescape(alloc, brief_s) catch continue;
        const detail = unescape(alloc, detail_s) catch continue;
        items.append(alloc, .{
            .id = id,
            .status = status,
            .name = name,
            .brief = brief,
            .detail = detail,
        }) catch return null;
    }
    // Recount statuses from the kept items so the UI header can never disagree
    // with the painted list after the MAX_ITEMS cap (review Minor).
    var ok: u32 = 0;
    var fail: u32 = 0;
    var pending: u32 = 0;
    for (items.items) |it| {
        switch (it.status) {
            .ok => ok += 1,
            .fail => fail += 1,
            .running => pending += 1,
        }
    }
    const run = ToolRun{
        .ok = ok,
        .fail = fail,
        .pending = pending,
        .items = items.toOwnedSlice(alloc) catch return null,
    };
    return .{ .run = run, .alloc = alloc };
}

test "decode round-trip with escaping" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    // Header counts are recounted from the kept items (review Minor), so even a
    // lying wire header `9/9/9` can never disagree with the painted items.
    const text =
        "toolrun\t1\t9/9/9\n" ++
        "1\tok\tread_file\tread_file lib/x.ts ok\tread_file · ✓ ok · lib/x.ts · 3 lines\n" ++
        "2\tfail\texec\t\texec · ✗ failed · exit=1\n" ++
        "3\trunning\tcat\tcat · running…\t";
    var d = decode(a, text) orelse return error.ExpectedDecode;
    defer d.deinit();

    try std.testing.expectEqual(@as(u32, 1), d.run.ok);
    try std.testing.expectEqual(@as(u32, 1), d.run.fail);
    try std.testing.expectEqual(@as(u32, 1), d.run.pending);
    try std.testing.expectEqual(@as(usize, 3), d.run.items.len);

    try std.testing.expectEqual(@as(u32, 1), d.run.items[0].id);
    try std.testing.expectEqual(Status.ok, d.run.items[0].status);
    try std.testing.expectEqualStrings("read_file", d.run.items[0].name);
    try std.testing.expectEqualStrings("read_file · ✓ ok · lib/x.ts · 3 lines", d.run.items[0].detail);

    try std.testing.expectEqual(Status.fail, d.run.items[1].status);
    try std.testing.expectEqual(Status.running, d.run.items[2].status);
    try std.testing.expectEqualStrings("cat · running…", d.run.items[2].brief);
}

test "decode escapes tab/newline/backslash" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const text = "toolrun\t1\t0/0/1\n" ++
        "1\tok\twith\\ttab\tline one\\nline two\tback\\\\slash\\tend";
    var d = decode(a, text) orelse return error.ExpectedDecode;
    defer d.deinit();
    try std.testing.expectEqualStrings("with\ttab", d.run.items[0].name);
    try std.testing.expectEqualStrings("line one\nline two", d.run.items[0].brief);
    try std.testing.expectEqualStrings("back\\slash\tend", d.run.items[0].detail);
}

test "decode fails open on bad header / unknown version" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    try std.testing.expect(decode(a, "") == null);
    try std.testing.expect(decode(a, "garbage") == null);
    try std.testing.expect(decode(a, "toolrun\t99\t1/0/0") == null);
    try std.testing.expect(decode(a, "toolrun\tjunk\t1/0/0") == null);
    try std.testing.expect(decode(a, "notatoolrun\t1\t1/0/0") == null);
}

test "decode tolerates malformed item lines" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const text = "toolrun\t1\t1/0/0\n" ++
        "1\tok\tread_file\tbrief\tdetail\n" ++
        "bad.line" ++
        "\n2\twat\tbad\n";
    var d = decode(a, text) orelse return error.ExpectedDecode;
    defer d.deinit();
    try std.testing.expectEqual(@as(usize, 1), d.run.items.len);
    try std.testing.expectEqualStrings("read_file", d.run.items[0].name);
}

test "ui id packing is unique for a full group and across rows" {
    // Mirrors the IMGUI id scheme in `ui/toolrun.zig` paintToolRun: each item owns a
    // `item_stride`-wide namespace (`it_id * item_stride + slot`) under a row
    // that advances by `row_step` = msg_index *% 1000003. Serves as the
    // reviewer-requested cheap guard that MAX_ITEMS items never alias.
    const item_stride: usize = 1024;
    const row_step: usize = 1000003;
    const slot_max: usize = 4;

    // Within one row: every (item, slot) pair maps to a distinct widget id.
    var seen = std.AutoHashMap(u64, void).init(std.testing.allocator);
    defer seen.deinit();
    var it_id: usize = 1;
    while (it_id <= MAX_ITEMS) : (it_id += 1) {
        var slot: usize = 0;
        while (slot <= slot_max) : (slot += 1) {
            const id: u64 = @as(u64, it_id) * item_stride + slot;
            // fetchPut returns null when the key is new — a non-null result
            // means two (item, slot) pairs aliased.
            try std.testing.expect(try seen.fetchPut(id, {}) == null);
        }
    }
    // A full 200-item group's highest widget id stays below the next row step,
    // so two tool-run rows (distinct msg_index) never overlap widget id space.
    const group_span: u64 = @as(u64, MAX_ITEMS) * item_stride + slot_max;
    try std.testing.expect(group_span < row_step);
}

test "decode caps items at MAX_ITEMS (defense against dense/restored blobs)" {
    var arena = std.heap.ArenaAllocator.init(std.testing.allocator);
    defer arena.deinit();
    const a = arena.allocator();

    const ta = a;
    var text = std.ArrayList(u8).empty;
    var line_buf: [128]u8 = undefined;
    const head = std.fmt.bufPrint(
        &line_buf,
        "toolrun\t{d}\t{d}/0/0\n",
        .{ TOOL_RUN_VERSION, MAX_ITEMS + 5 },
    ) catch unreachable;
    try text.appendSlice(ta, head);
    var i: u32 = 0;
    while (i < MAX_ITEMS + 5) : (i += 1) {
        const line = std.fmt.bufPrint(
            &line_buf,
            "{d}\tok\tt{d}\tbrief\tdetail\n",
            .{ i + 1, i },
        ) catch unreachable;
        try text.appendSlice(ta, line);
    }

    var d = decode(a, text.items) orelse return error.ExpectedDecode;
    defer d.deinit();
    try std.testing.expectEqual(@as(usize, MAX_ITEMS), d.run.items.len);
}
